### ISSUE TITLE
listpays rpc method could expose the payment_hash if the bolt11 is null (keysend)

### ISSUE THREAD

**Issue by vincenzopalazzo** [ElementsProject/lightning#3880]
Hi all.

During my work on introducing a new command `delpay` (good fist issue), I noted that after the evolution of pay plugin, the method RPC `listpays` could return payment without `id`, such as the `bolt11`  value null if is used keysend plugin.

In cases of `bolt11` value null, I can use the command `listsendpays` to have a result with more details. However, if I don't have a fault on my idea the command `listpays` could be able to return the `bolt11` in cases of payment with invoice and `payment_hash` in cases of payment without invoice. 

An example of output could be the json below

### keysend payment

```json
{
      "payment_hash": "7cdf931c7a23efb6815ff239517ef6d373bb8d040caefba96314368afd93252d",
      "status": "complete",
      "preimage": "9a4f8576e0380394ab791b35e08793533b7d1d3303c383d91fda14a13978e7db",
      "amount_sent_msat": "1000msat"
  },
```

### invoice payment

```json
      {
         "bolt11": "lntb10n1p03cskapp5f730rvqpqelvqmtljkuxjku2ek00qnqmf5g3pcaefc06q6rmk8sqdp9ve5hsatsta3x7mr5xyckuatvd30hgam9d3mx2xqyjw5qcqp2sp5nh2swvkuup2qrekvlmstswh8kdjjh7chyuelp7sgy0hvtz9dzhqqrzjqdsdeghn2vmdxqmy8flmzu46vxzmjzr258aavp36z3rs2redmg8cwxe0fyqqq2qqqqqqqqqpqqqqqzsqqc9qy9qsqu7l2057sh8zvpw56ftu87semua2a6a4ftjscvvfewz0gj92dsylx7whv30heqy0zr6jy6udtad57m93q3pmyezfcf4pntjqstks0v5qpedegh0",
         "status": "complete",
         "preimage": "7516ea2e790e5e8edb02709682e445d38aea07c7e772db9441b5091908d19c23",
         "amount_sent_msat": "1000msat"
      }
```

The content above make sense? or use a different json format inside the same rpc command and it is not correct?

reference implementation ~https://github.com/vincenzopalazzo/lightning/commit/7c13b0be79ff659597e345ad202b6fb692c5f57a~
https://github.com/vincenzopalazzo/lightning/commit/b59f72b201cde197f4b94e53e8d4df1f44348451

**Comment by vincenzopalazzo** [comment]
Also, I discovered my fault, in the past, there was a case that the bolt11 was null because these values did not store inside DB. In this case, the json response of the `listpays` returned a different format that it included also `payment_hash`.

The code that built a different format if the bol11 was null (in 0.7.3) is reported below

https://github.com/ElementsProject/lightning/blob/5bc2de89979ca8eebb7b4b05da122aeb342c6f31/plugins/pay.c#L1248-L1257

With the actual code, I'm not able to see these code or a code that do this behavior, (maybe is my fault or maybe this retro compatibility was break and for these reason, we see in my original post the bolt11 value null).

In conclusion, the command `listpay` is the command which is possible to maintain the compatibility with new and old pay. For instance, with keysend and pay, but in this cases, if I have the bolt11 null and I call the command `listpays bolt11` I don't receive the correct behavior (maybe I can receive also a crash?)

For this reason, it could to have the possibility to pass a new value inside the `listpay` command such as `payment_hash`.

**Comment by cdecker** [comment]
I think you're right, we should add the `payment_hash` to the results you mentioned, potentially always, even if we have a `bolt11` field, since it's about the only field that is guaranteed to be there.

**Comment by vincenzopalazzo** [comment]
I implemented somethings, I can open a PR, but before I want make enough tests and make all changes inside the docs file. To make a good work

**Comment by ZmnSCPxj** [comment]
fixed by #3888
